### PR TITLE
Support for multiple joystick types

### PR DESCRIPTION
**Pull request by yabberyabber**
I'm just creating this PR so that I can comment online. Feel free to keep working on this until you think it's ready for an actual PR

### CODE REVIEW

File: src/TeleopMode.cpp [hunk 1]
```diff
@@ -1,20 +1,20 @@
-/*
- * TeleopMode.cpp
- *
- *  Created on: January 7, 2018
- *      Authors: Kyle, Chris
- *
- *  Control map available at: https://goo.gl/MrViHA
- */
+/*m_driverjoystick*/ /*
```
Comment (yabberyabber) [inline src/TeleopMode.cpp]: was `/*m_driverjoystick*/` added by accident?

File: src/TeleopMode.cpp [hunk 2]
```diff
@@ -141,17 +158,106 @@ void Teleop::TeleopPeriodic() {
...
     else {
         DBStringPrintf(DBStringPos::DB_LINE2, "closed loop wrist");
-    }
+    }*/
 }
 
 void Teleop::TeleopStop() {
 }
 
+void Teleop::HandlePoofsJoystick(uint32_t port, uint32_t button,
```
Comment (yabberyabber) [inline src/TeleopMode.cpp]: Do we want to have a different handler call for each joystick?  The answer might be yes (I'll trust your judgement) but I want to make sure you've thought through the pro's and cons.  The other option would be for them all to call the same `HandleJoystick` function

File: lib/helpers/PoofsJoystickHelper.h [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    /**
+     * Get the value of the given axis with deadband.
+     *
+     * @param axis Specifies the axis to get the value of.
+     * @param fSquared Specifies whether the joystick input should be squared.
+     * @param threshold Specifies the deadband threshold.
+     * @param hand Specifies the handedness of the joystick (default to right
+     *        hand).
+     */
+    float GetRawAxisWithDeadband(int axis, bool fSquared = false,
```
Comment (yabberyabber) [inline lib/helpers/PoofsJoystickHelper.h]: This function is the same in all these classes.  Sounds like a GREYT opportunity for inheritance ;)

**Comment by yabberyabber** [comment]
originally we had the following classes:

 - JoystickObserver
    - this class only has one function... observeJoystickStateChange
    - In our codebase we have one class that inherits from JoystickObserver.  That class is Robot.  When robot makes an ObservableJoystick, it passes `this` as a parameter because the Robot class wants to get notified whenever a button gets pressed or released

 - ObservableJoystick
    - this class is just like a joystick except whenever a button is pressed or released, it calls observeJoystickStateChange

These two classes need to be separate

If you want, you can have a pair of classes for each joystick type.  You could have:

 - for the poofs joysticks:
    - PoofsJoystickObserver which has an `observePoofsJoystickStateChange` function (Robot would inherit from this and would define observePoofsJoystickStateChange)
    - ObservablePoofsJoystick which calls `observePoofsJoystickStateChange`
 - for dual action joysticks
    - DualActionJoystickObserver which has an `observeDualActionJoystickStateChange` function (Robot would inherit from this etc)
    - ObservableDualActionJoystick which calls `observeDualActionJoystickStateChange`
 - and so on


Another approach would be to have only one JoystickObserver class.  the function `observeDualActionJoystickStateChange` would take 4 arguments: `port`, `buttonId`, `pressedP`, and `joystickType` where joystickType would say which kind of joystick it is.

there would still be a different ObservableXXXJoystick class for each joystick type 
 - ObservablePoofsJoystick would call `observeJoystickStateChange` and pass like "poofs" or something so you can tell it's from the poofs class
 - ObservableDualActionJoystick would call `observeJoystickStateChange` and pass like "dualAction" or something so you can tell it's the dualAction class





Either way, I suggest you ignore my advice about giving the observable joystick classes a common parent class at first.  Get it working first and go back and do the parent class later

**Comment by yabberyabber** [comment]
The pattern we're going for here is called the observer pattern.  It makes things faster because Robot doesn't constantly ask over and over again "is a pressed? is b pressed? is x pressed? etc..."  instead, Joystick tells Robot only when things change.  https://medium.com/@gayashanbc/explore-the-observer-design-pattern-with-c-8592459b0486  here's a good article on the background and rational, but in our case we only have one observer and it's called Robot

File: lib/helpers/PoofsJoystickHelper.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    if (fSquared) {
+        value = Util::signSquare(value);
+    }
+
+    return value;
+}
+
+bool ObservablePoofsJoystick::GetLTrigger() {
+    if (this->GetRawButton(PoofsJoysticks::LeftTrigger)) {
+        m_lastLTrigger = true;
```
Comment (yabberyabber) [inline lib/helpers/PoofsJoystickHelper.cpp]: this entire function could be

```
bool ObservablePoofsJoystick::GetLTrigger() {
    return this->GetRawButton(PoofsJoysticks::LeftTrigger);
}
```

File: lib/helpers/PoofsJoystickHelper.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    return m_lastLTrigger;
+}
+
+bool ObservablePoofsJoystick::GetLBumper() {
+    if (this->GetRawButton(PoofsJoysticks::LeftBumper)) {
+        m_lastLBumper = true;
+    }
+    else {
```
Comment (yabberyabber) [inline lib/helpers/PoofsJoystickHelper.cpp]: Same applies for most of these

File: lib/helpers/PoofsJoystickHelper.h [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+using namespace frc;
+
+namespace frc973 {
+
+namespace PoofsJoysticks {
+const unsigned int LeftXAxis = 1;
+const unsigned int LeftYAxis = 0;
+const unsigned int RightXAxis = 3;
+const unsigned int RightYAxis = 2;
```
Comment (yabberyabber) [inline lib/helpers/PoofsJoystickHelper.h]: nice job with the separate button namespaces

File: src/TeleopMode.cpp [hunk 7]
```diff
@@ -98,20 +106,39 @@ void Teleop::TeleopPeriodic() {
...
                 m_cubeIntakeState = CubeIntakeState::Idle;
             }
             break;
         case CubeIntakeState::ManualIntaking:
             m_intakeAssembly->RunIntake(-1.0);
             if (m_intakeAssembly->GetClaw()->IsCubeIn()) {
-                m_cubeIntakeState = CubeIntakeState::Idle;
+                m_operatorJoystick->SetRumble(
+                    GenericHID::RumbleType::kLeftRumble, 1);
+                m_operatorJoystick->SetRumble(
```
Comment (yabberyabber) [inline src/TeleopMode.cpp]: This is dope

**Comment by Chris2fourlaw** [comment]
Heh. Should have seen Adam’s face when we enabled rumble mode.   
  

  
  

  
  
>

File: src/TeleopMode.cpp (hunk 7 above)
Comment (KyleD973) [reply src/TeleopMode.cpp]: Rumble and all joysticks worked pretty fine! We should be using this at MTTD.

**Pull request merged**